File: DataScienceFromScratch/julia/4.jl
#!/usr/local/bin/julia

######################################### Vectors #############################################
hght_w8_age = [70, #inches
              170, #pounds
              40] #years
grades = [95, #exam1
          80, #exam2
          75, #exam3
          62] #exam4

vector_add(v,w) = [v+w for (v,w) in zip(v,w)]
vector_subtract(v,w) = [v-w for (v,w) in zip(v,w)]
vectorsum(vectors) = reduce(vector_add,vectors)

#component wise mean
vector_mean(vectors) = mul!(Y=similar(vectors), 1/(length(vectors)), vectorsum(vectors))

#v1*v1 + v2*v2 +...+ vn*wn
sum_of_squares(v) = dot(v,v) 

magnitude(v) = sqrt(sum_of_squares(v))

#distance between two vectors
squared_distance(v,w) = sum_of_squares(vector_subtract(v,w))
distance(v,w) = magnitude(vector_subtract(v,w))


######################################### Matrices ############################################
#2D collection of numbers: lists of lists
B = [1 2
     3 4 
     5 6]

A = [1 2 3
     4 5 6]

Ap = reshape(1:6, 3,2)' #julia is column major by default. ' symbol adjoins a matrix
Bp = reshape(1:6, 2,3)' #trick to get matrices axis flipped

println("A: $A\nAp: $Ap\nB: $B\nBp: $Bp")
println("A == Ap?: $(A==Ap)\nB == Bp?: $(B==Bp)")

println("A (ndims, size, length, row 1, column 1): $(ndims(Ap)), $(size(Ap)), $(length(Ap)),",
        " $(getindex(Ap, 1:2:6)) or $(first(eachrow(Ap))),",
        " $(getindex(Ap, 1:2)) or $(first(eachcol(Ap)))",
        "B (ndims, size, length): $(ndims(Bp)), $(size(Bp)), $(length(Bp))\n",
        "B (row 1, column 1): $(getindex(Bp, 1:3:6)), $(getindex(Bp, 1:3))\n")

using LinearAlgebra
@show Matrix(1*I,5,5) # identity matrix
"""
  IRL: use LinearAlgebra library, this was just for intro
"""

File: DataScienceFromScratch/python/4.py
#!/usr/local/bin/python2.7

height_weight_age = [70, # inches
                    170, # pounds
                    40 ] # years

grades = [95,# exam1
          80,# exam2
          75,# exam3
          62 ] # exam4

def vector_add(v, w):
    """adds corresponding elements""" 
    return [v_i + w_i for v_i, w_i in zip(v, w)]

def vector_subtract(v, w):
    """subtracts corresponding elements""" 
    return [v_i - w_i for v_i, w_i in zip(v, w)]

def vector_sum(vectors):
    """sums all corresponding elements""" 
    result = vectors[0]                     #start with first vec
    for vector in vectors[1:]:              #then loop over rest
        result = vector_add(result, vector) #add to result
    return result

def vectorsum(vectors):
    return reduce(vector_add, vectors)

def scalar_multiply(c, v):
    """c is a number, v is a vector""" 
    return [c * v_i for v_i in v]

def vector_mean(vectors):
    """compute the vector whose ith element is the mean of the 
    ith elements of the input vectors"""
    n = len(vectors)
    return scalar_multiply(1/n, vector_sum(vectors))


def dot(v, w):
    """v_1 * w_1 + ... + v_n * w_n""" 
    return sum(v_i * w_i for v_i, w_i in zip(v, w))

def sum_of_squares(v):
    """v_1 * v_1 + ... + v_n * v_n""" 
    return dot(v, v)

import math
def magnitude(v):
    return math.sqrt(sum_of_squares(v))

def squared_distance(v, w):
    """(v_1 - w_1) ** 2 + ... + (v_n - w_n) ** 2""" 
    return sum_of_squares(vector_subtract(v, w))

def distance(v, w):
    return magnitude(vector_subtract(v, w))

A = [[1,2,3], #A has 2 rows and 3 columns 
    [4, 5, 6]]

B = [[1, 2], # B has 3 rows and 2 columns 
    [3, 4],
    [5, 6]]


def shape(A):
    num_rows = len(A)
    num_cols = len(A[0]) if A else 0 # number of elements in first row 
    return num_rows, num_cols

def get_row(A, i):
    return A[i] # A[i] is already the ith row

def get_column(A, j): # jth element of row A_i
    return [A_i[j] for A_i in A] # for each row A_i

def make_matrix(num_rows, num_cols, entry_fn):
    """returns a num_rows x num_cols matrix whose (i,j)th entry is entry_fn(i, j)"""
    return [[entry_fn(i, j)  # given i, create a list
            for j in range(num_cols)] # [entry_fn(i, 0), ... ] 
            for i in range(num_rows)] # create one list for each

def is_diagonal(i, j):
    """1's on the 'diagonal', 0's everywhere else"""
    return 1 if i == j else 0 

identity_matrix = make_matrix(5, 5, is_diagonal)
print A, B, identity_matrix



